### ISSUE TITLE
Gradients greatly change after `BetterTransformer.transform` application

### ISSUE THREAD

**Issue by lengstrom** [huggingface/optimum#1091]
### System Info

```shell
optimum: 1.8.6
transformers: 4.29.2
ubuntu/CUDA 11.7/pytorch 2.1
```

### Who can help?

@fxmarty @younesbelkada

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

https://gist.github.com/lengstrom/84a5d0d95f8942eb4eca82d9eb5aa272

This code should perfectly reproduce with just pytorch 2.1, optimum, datasets, numpy, and transformers installed. The output is (full log here: https://gist.github.com/lengstrom/84a5d0d95f8942eb4eca82d9eb5aa272?permalink_comment_id=4593821#gistcomment-4593821):
```
tensor([0.9919, 0.9719, 0.9756, 0.9666, 0.9827, 0.9827, 0.9839, 0.9857, 0.9909,
        0.9564, 0.9752, 0.9904, 0.9798, 0.9764, 0.9827, 0.9842])
cosine sim for gpt_neox.embed_in.weight:  tensor(0.0670) tensor(0.3000)
cosine sim for gpt_neox.layers.0.input_layernorm.weight:  tensor(0.7921) tensor(0.1659)
cosine sim for gpt_neox.layers.0.input_layernorm.bias:  tensor(0.8910) tensor(0.1325)
cosine sim for gpt_neox.layers.0.post_attention_layernorm.weight:  tensor(0.8930) tensor(0.0971)
```

This log shows that the gradients are often rather different, and that in some parameter groups the compared gradients are nearly *uncorrelated* (i.e., close to 0)

### Expected behavior

The gradients of samples for models with and without `BetterTransformer` transformation should be nearly identical, but are in practice often very different.

Considering a GPTNeoX model, we fix a sample and record the gradient of the BetterTransformer-transformed model and the gradient of the original model. We then measure the cosine similarity (a vector similarity metric, 0 = uncorrelated, 1 = perfectly correlated) and find that both:

- (a) the cosine similarity of gradients between BetterTransform and standard models is not 1.0 and
- (b) on some parameter groups, this cosine similarity is near 0 (i.e., almost uncorrelated!) - see above the entry for `gpt_neox.embed_in.weight`

Is there a reason for this behavior? Based on the documentation, the only difference between the transformed and untransformed models is different code paths being used, which should not change the underlying gradients.

**Comment by lengstrom** [comment]
Intriguingly, the logits (i.e., just the result of the forward pass) are super close to one another:
```
ch.nn.functional.cosine_similarity(logits.reshape(bs, -1), logits_btfm.reshape(bs, -1), dim=1)
```
yields
```
tensor([1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000,
        1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000, 1.0000],
       device='cuda:0', grad_fn=<SumBackward1>)
```

**Comment by fxmarty** [comment]
Thank you - this is rather critical, apologies this should have been tested upfront. I had a try reproducing with just `torch.nn.functional.scaled_dot_product_attention` and the boiler plate code from [PyTorch doc](https://pytorch.org/docs/master/generated/torch.nn.functional.scaled_dot_product_attention.html?highlight=scaled_dot_product_attention#torch.nn.functional.scaled_dot_product_attention) - gradients indeed match there. So the issue is really in our implementation. Let me have a look, debug and add tests for this.

**Comment by fxmarty** [comment]
I can't reproduce the issue with gpt2 and gpt-neo type models, but can with gpt-neox.

Here's a simplier script (yours does not work when I change the model):
```python
from transformers import AutoTokenizer, AutoModelForCausalLM, AutoConfig, GPT2LMHeadModel
from optimum.bettertransformer import BetterTransformer

import torch
import copy

import functools


def recurse_getattr(obj, attr: str):
    """
    Recursive `getattr`.

    Args:
        obj:
            A class instance holding the attribute.
        attr (`str`):
            The attribute that is to be retrieved, e.g. 'attribute1.attribute2'.
    """

    def _getattr(obj, attr):
        return getattr(obj, attr)

    return functools.reduce(_getattr, [obj] + attr.split("."))

#model_id = "hf-internal-testing/tiny-random-gpt2"

"""
model_id = "gpt2-xl"
cfg = AutoConfig.from_pretrained(model_id)
cfg.embd_pdrop = 0
cfg.attn_pdrop = 0
cfg.resid_pdrop = 0

model_vanilla = GPT2LMHeadModel(cfg)
model_sdpa = copy.deepcopy(model_vanilla)
"""
#model_id = "hf-internal-testing/tiny-random-GPTNeoXForCausalLM"

model_id = "EleutherAI/pythia-160m"
#model_id = "EleutherAI/gpt-neo-125m"
model_vanilla = AutoModelForCausalLM.from_pretrained(model_id)
model_sdpa = AutoModelForCausalLM.from_pretrained(model_id)

tokenizer = AutoTokenizer.from_pretrained(model_id)

model_sdpa = BetterTransformer.transform(model_sdpa)

inp = tokenizer("This is just to test gradients", return_tensors="pt")

model_vanilla = model_vanilla.train()
model_sdpa = model_sdpa.train()
model_vanilla.zero_grad()
model_sdpa.zero_grad()

#model_vanilla = model_vanilla.eval()
#model_sdpa = model_sdpa.eval()

#with torch.inference_mode():
res_vanilla = model_vanilla(**inp).logits

print("\n\n\n\n RUNNING SDPA")
res_sdpa = model_sdpa(**inp).logits


print("Res match:", torch.allclose(res_vanilla, res_sdpa), "Maxdiff:", torch.max(torch.abs(res_vanilla - res_sdpa)).item())

loss_vanilla = res_vanilla.sum()
loss_sdpa = res_sdpa.sum()

loss_vanilla.backward()
loss_sdpa.backward()

print("Loss match:", torch.allclose(loss_vanilla, loss_sdpa), "Maxdiff:", torch.max(torch.abs(loss_vanilla - loss_sdpa)).item())

for name, param_vanilla in model_vanilla.named_parameters():
    param_sdpa = recurse_getattr(model_sdpa, name)

    maxdiff = torch.max(torch.abs(param_vanilla.grad - param_sdpa.grad)).item()
    relativediff = torch.mean(torch.abs(param_vanilla.grad - param_sdpa.grad) / torch.abs(param_vanilla.grad))
    print(f"{name} grad match:", torch.allclose(param_vanilla.grad, param_sdpa.grad), f"Maxdiff: {maxdiff}, relativediff: {relativediff}")
```

**Comment by fxmarty** [comment]
@lengstrom I changed nothing. I used a simplier script (tried to change the model on yours but got into errors) and can not reproduce the issue. Do you see anything wrong with my script? A difference is the device, I did not try on cuda (which may in turn use the flash attention or memory efficient attenton kernel).

**Comment by lengstrom** [comment]
@fxmarty : changing three basic things in your script I am able to replicate the behavior I showed.

1. Put model / inputs on GPU (I have an 80GB A100)
2. Use the 70m model instead of 160m model
3. 3x longer input4. 

Not doing any of these three things will make you miss the behavior I showed. Furthermore, I found that changing the loss function to cross entropy instead of the mean of the output logits also exacerbates the issue. Full script:

```
from transformers import AutoTokenizer, AutoModelForCausalLM, AutoConfig, GPT2LMHeadModel
from optimum.bettertransformer import BetterTransformer

import torch
import copy
import functools


def recurse_getattr(obj, attr: str):
    def _getattr(obj, attr):
        return getattr(obj, attr)

    return functools.reduce(_getattr, [obj] + attr.split("."))


model_id = "EleutherAI/pythia-70m"
model_vanilla = AutoModelForCausalLM.from_pretrained(model_id)
model_sdpa = AutoModelForCausalLM.from_pretrained(model_id)

for name, param_vanilla in model_vanilla.named_parameters():
    param_sdpa = recurse_getattr(model_sdpa, name)

    maxdiff = torch.max(torch.abs(param_vanilla - param_sdpa)).item()
    relativediff = torch.mean(torch.abs(param_vanilla - param_sdpa) / torch.abs(param_vanilla))
    print(f"{name} param match:", torch.allclose(param_vanilla, param_sdpa), f"Maxdiff: {maxdiff}, relativediff: {relativediff}")


tokenizer = AutoTokenizer.from_pretrained(model_id)

model_sdpa = BetterTransformer.transform(model_sdpa)

inp = tokenizer(["This is just to test gradients. Thisi s just to test gradients? This is just to test gradients!"] * 1, return_tensors="pt")
# make inp vectors on gpu
inp = {k: v.cuda() for k, v in inp.items()}

model_vanilla = model_vanilla.eval().cuda()
model_sdpa = model_sdpa.eval().cuda()
model_vanilla.zero_grad()
model_sdpa.zero_grad()

model_vanilla = model_vanilla.eval()
model_sdpa = model_sdpa.eval()

torch.set_printoptions(threshold=100000000000)

res_vanilla = model_vanilla(**inp).logits

print("\n\n\n\n RUNNING SDPA")
res_sdpa = model_sdpa(**inp).logits


print("Res match:", torch.allclose(res_vanilla, res_sdpa), "Maxdiff:", torch.max(torch.abs(res_vanilla - res_sdpa)).item())

# loss_vanilla = res_vanilla.mean()
# loss_sdpa = res_sdpa.mean()
loss_vanilla = torch.nn.functional.cross_entropy(res_vanilla.view(-1, res_vanilla.shape[-1]), inp["input_ids"].view(-1)).mean()
loss_sdpa = torch.nn.functional.cross_entropy(res_sdpa.view(-1, res_sdpa.shape[-1]), inp["input_ids"].view(-1)).mean()

loss_vanilla.backward()
loss_sdpa.backward()

print("Loss match:", torch.allclose(loss_vanilla, loss_sdpa), "Maxdiff:", torch.max(torch.abs(loss_vanilla - loss_sdpa)).item())

for name, param_vanilla in model_vanilla.named_parameters():
    param_sdpa = recurse_getattr(model_sdpa, name)

    maxdiff = torch.max(torch.abs(param_vanilla.grad - param_sdpa.grad)).item()
    relativediff = torch.mean(torch.abs(param_vanilla.grad - param_sdpa.grad) / torch.abs(param_vanilla.grad))
    cosine = torch.nn.functional.cosine_similarity(param_vanilla.grad.flatten(), param_sdpa.grad.flatten(), dim=0)
    print(f"{name} grad match:", torch.allclose(param_vanilla.grad, param_sdpa.grad), f"Maxdiff: {maxdiff}, relativediff: {relativediff}, cosine={cosine}")
```

**Comment by lengstrom** [comment]
It looks like memory efficient SDP is the issue - when I disable it the script works again:

```
torch.backends.cuda.enable_mem_efficient_sdp(False)
torch.backends.cuda.enable_flash_sdp(False)
torch.backends.cuda.enable_math_sdp(True)
```

**Comment by fxmarty** [comment]
Thank you for investigating it.

**Comment by lengstrom** [comment]
interesting, thanks for looking into it more! if it is informative: using the Flash SDP I found the same issue with the 70m model (albeit - I had to do a lot of float/half casting to get the NeoX model to even run without errors).

**Comment by fxmarty** [comment]
I can actually only reproduce with mem-efficient kernel (flash fp16 is fine to me):
![image](https://github.com/huggingface/optimum/assets/9808326/f6801661-1707-419e-8623-23070a45f743)

**Comment by lengstrom** [comment]
cool! this is not what I found - but I compared the gradients of vanilla fp32 models with flash attention fp16 models so that could be why. sleeping now but will take a look tomorrow..

**Comment by lengstrom** [comment]
curiosity got the better of me, its weird that you get consistent half prec performance - when I modify both models to use half precision in the original script from you that i modified above I get the same kinds of cosine similarity issues: (regardless of whether or not the vanilla model is fp16 or fp32) - https://gist.github.com/lengstrom/39267bbeb9a40ec68f21e71fb60b1b2b

**Comment by fxmarty** [comment]
My bad - was just an error in copy/paste on the plot:
![image](https://github.com/huggingface/optimum/assets/9808326/0a7b4e15-854f-44e0-baf5-260fa7a35b07)
